### ISSUE TITLE
Tree shaking doesn't work on second-level-deep modules /with example

### ISSUE THREAD

**Issue by fvigotti** [webpack/webpack#11194]
# Bug report

**What is the current behavior?**
simple webpack tree shaking setup with typescript correctly works on 1st level nested functions, but on second-level nested functions doesn't remove unused/never-called code 

full working example with readme containing the issue is here : 
https://github.com/fvigotti/tree-shaking-test1/tree/master

**If the current behavior is a bug, please provide the steps to reproduce.**
`npx webpack --mode=production`

**What is the expected behavior?**
the built minified file `dist/entry3-call-sub.js` should not contain those functions because never called 
```js
function u(e) {
        console.log("module 1 - b called: " + e);
    }
    function r(e) {
        console.log("module 1 - c called: " + e);
    }
function u(e) {
        console.log("module 2 - b called: " + e);
    }
    function r(e) {
        console.log("module 2 - c called: " + e);
    }

```

**Other relevant information:**
webpack version: ^4.43.0

**Comment by vankop** [comment]
Should work with `webpack@next`. Unfortunately, this feature implemented only in webpack@5.

**Comment by evilebottnawi** [comment]
Answer above

**Comment by fvigotti** [comment]
well.. no.. same problem in webpack 5

https://github.com/fvigotti/tree-shaking-test1/tree/webpacknext

**Comment by fvigotti** [comment]
one thing that probably could give hints about the internal issue is that seems that dependencies are not entry-point independent.. I mean: if another entry point call a function this is not tree-shaked in other entry points..
but the logic of multiple entry points is that those should be independent!

ie:
if you disable this entry in webpack.config.js:
```
//  "entry2-call-something" : "./src/entry2-call-something.ts",
```
the entry3 works.. but this behaviour is unwanted and imo a bug..

**Comment by vankop** [comment]
@fvigotti https://github.com/webpack/changelog-v5#inner-module-tree-shaking

**Comment by fvigotti** [comment]
@vankop 
thanks I see now that in the description of the webpack 5 tree shaking it says that webpack 4 does not work on sub-modules, but by definition the term tree-shaking should be a cherry picking of the necessary dependencies(classes/functions/..) and it is implied that it also works for sub-modules,
if this is not the case in the webpack 4 version, it should be specified in the webpack 4 guide and not in the webpack 5 guide (saying that webpack4 doesn't work) , because practically having a tree-shaking that works only for direct top-level dependencies is almost completely useless. and maybe someone like me that was looking for tree-shaking in typescript and started using webpack for that.. maybe will go elsewere if that feature is so much limited here.. 

again:
given this, however, the problem described remains that does not work even in webpack 5 for multiple entrypoints, I do not know how it is implemented but imo the dependency graph should be reconstructed for each entry point and maybe it is not so

**Comment by vankop** [comment]
> if this is not the case in the webpack 4 version, it should be specified in the webpack 4 guide 

Feel free to make a pull request https://github.com/webpack/webpack.js.org/tree/webpack-4

> the problem described remains that does not work even in webpack 5 for multiple entrypoints

Yes, this does not work yet and probably will be implemented in future as a new feature. The problem here, as I see, is that webpack does not know how you will execute your scripts. For example for web app you probably don't want initialize your module twice because of 2 reasons:
- performance, transfer code size
- how to handle modules with side-effect more smarter, e.g.
```
let counter = 1;
export const inc = () => counter++;
export const dec = () => counter++;
export const getCounter = () => counter;
```

So to sum up, there are some challenges.. and probably a lot of webpack internals need to be refactored..

**Comment by fvigotti** [comment]
plus, 
Gonig back to webpack 4, and disabling entry2 from my example, entry3 is correctly tree-shaked , so I see exactly same behaviour in this example from webpack4 and webpack5 with exactly same issue.. that now after more investigation i repeat is not related to second-level deep modules but more with dependencies-graph-interference  between entries ( webpack.config : entry : {..these..} )

**Comment by vankop** [comment]
Sorry, I misunderstood problem because of the title.. Yes, this is tree-shaking per entrypoint problem

**Comment by fvigotti** [comment]
Thank you for help, I still have some issues with three shaking but maybe it's material for stackoverflow..
given that you understand the issue and that the title for this issue is now incorrect I feel right to close the comment, you can open a new one and reference this if you need to keep track of the issue 
thank you,
Francesco

**Comment by sokra** [comment]
https://github.com/webpack/webpack/pull/11235 will fix your other problem once it's done.
